### PR TITLE
Add shutdown behaviour section to cancellation token upgrade guide

### PR DESCRIPTION
**Pull request by andreasohlund**
Replaces https://github.com/Particular/docs.particular.net/pull/5292

### CODE REVIEW

File: nservicebus/upgrades/7to8/index.md [hunk 1]
```diff
@@ -96,6 +96,10 @@ If cancellation is not a major concern, the [analyzer severity](https://docs.mic
 dotnet_diagnostic.NSB0002.severity = suggestion
 ```
 
+### Shutdown behavior
+
+In Version 7 MSMQ, RabbitMQ and SqlServer allowed handlers to take up to 30 seconds to finish before completing the shutdown. In Version 8 transports will block shutdown until all in-flights handlers complete. The cancellation token available on the message context in Version 8 will be triggered when the host aborts the shutdown and it's recommended that all handlers observes this token to enable graceful shutdown.
```
Comment (DavidBoike) [inline nservicebus/upgrades/7to8/index.md:101]: I added a lot of words. Do you like any of them?

```suggestion
In all versions of NServiceBus, there is a period after stopping the endpoint when NServiceBus will stop receiving new messages but give any currently running message handlers an opportunity to complete naturally.

In NServiceBus version 7 and below, the transports for MSMQ, RabbitMQ and SqlServer allowed handlers to take up to 30 seconds to finish before forcing an endpoint shutdown. In NServiceBus version 8, all transports will block shutdown until all in-flight handlers complete.

The cancellation token available on the `IMessageHandlerContext` in NServiceBus version 8 will be triggered when the host aborts the shutdown. For example, the .NET Generic Host signals the cancellation token after 5 seconds.

However, if handlers do not observe the cancellation token, for example if a handler called `Task.Delay(TimeSpan.Infinite)`, the endpoint will never be able to shut down.

Therefore, it is recommended that all message handlers observe the cancellation token to enable graceful shutdown.
```

File: nservicebus/upgrades/7to8/index.md [hunk 2]
```diff
@@ -96,6 +96,18 @@ If cancellation is not a major concern, the [analyzer severity](https://docs.mic
...
+
+In all versions of NServiceBus, there is a period after calling `endpoint.Stop()` when NServiceBus will stop receiving new messages but give any currently running message handlers an opportunity to complete naturally.
+
+In NServiceBus version 7 and below, the transports for MSMQ, RabbitMQ and SqlServer allowed handlers to take up to 30 seconds to finish before forcing an endpoint shutdown. In NServiceBus version 8, all transports will block shutdown until all in-flight handlers complete.
+
+The cancellation token available on the `IMessageHandlerContext` in NServiceBus version 8 will be triggered when the host aborts the shutdown. For example, the .NET Generic Host signals the cancellation token after 5 seconds.
+
+However, if handlers do not observe the cancellation token, for example if a handler called `Task.Delay(TimeSpan.Infinite)`, the endpoint will never be able to shut down.
+
+Therefore, it is recommended that all message handlers observe the cancellation token to enable graceful shutdown.
```
Comment (andreasohlund) [inline nservicebus/upgrades/7to8/index.md:109]: Perhaps we should say "long running message handlers"? Or would that water down the message?

File: nservicebus/upgrades/7to8/index.md [hunk 3]
```diff
@@ -96,6 +96,18 @@ If cancellation is not a major concern, the [analyzer severity](https://docs.mic
 dotnet_diagnostic.NSB0002.severity = suggestion
 ```
 
+### Shutdown behavior
+
+In all versions of NServiceBus, there is a period after calling `endpoint.Stop()` when NServiceBus will stop receiving new messages but give any currently running message handlers an opportunity to complete naturally.
```
Comment (adamralph) [inline nservicebus/upgrades/7to8/index.md:101]:
```suggestion
In all versions of NServiceBus, `endpoint.Stop()` immediately stops receiving new messages but waits for a period of time for currently running message handlers to complete.
```

File: nservicebus/upgrades/7to8/index.md [hunk 4]
```diff
@@ -96,6 +96,18 @@ If cancellation is not a major concern, the [analyzer severity](https://docs.mic
...
+
+In NServiceBus version 7 and below, the transports for MSMQ, RabbitMQ and SqlServer allowed handlers to take up to 30 seconds to finish before forcing an endpoint shutdown. In NServiceBus version 8, all transports will block shutdown until all in-flight handlers complete.
```
Comment (adamralph) [inline nservicebus/upgrades/7to8/index.md:103]:
```suggestion
In NServiceBus version 7 and below, the MSMQ, RabbitMQ, and SqlServer transports allow handlers up to 30 seconds to complete before forcing an endpoint shutdown. In NServiceBus version 8, all transports block shutdown until all handlers complete.
```

File: nservicebus/upgrades/7to8/index.md [hunk 5]
```diff
@@ -96,6 +96,18 @@ If cancellation is not a major concern, the [analyzer severity](https://docs.mic
...
+
+The cancellation token available on the `IMessageHandlerContext` in NServiceBus version 8 will be triggered when the host aborts the shutdown. For example, the .NET Generic Host signals the cancellation token after 5 seconds.
```
Comment (adamralph) [inline nservicebus/upgrades/7to8/index.md:105]:
```suggestion
The cancellation token available on the `IMessageHandlerContext` in NServiceBus version 8 is triggered when the host forces shutdown. For example, the .NET Generic Host signals the cancellation token after 5 seconds.
```

File: nservicebus/upgrades/7to8/index.md [hunk 6]
```diff
@@ -96,6 +96,18 @@ If cancellation is not a major concern, the [analyzer severity](https://docs.mic
...
+
+However, if handlers do not observe the cancellation token, for example if a handler called `Task.Delay(TimeSpan.Infinite)`, the endpoint will never be able to shut down.
```
Comment (adamralph) [inline nservicebus/upgrades/7to8/index.md:107]:
```suggestion
However, if handlers do not observe the cancellation token, they will complete before the endpoints shuts down. This means that if, for example, a handler calls `Task.Delay(TimeSpan.Infinite)`, the endpoint will never shut down.
```

File: nservicebus/upgrades/7to8/index.md (hunk 2 above)
Comment (adamralph) [reply nservicebus/upgrades/7to8/index.md:109]: I think long/short running is irrelevant, and those terms are anyway subjective.

```suggestion
Therefore, it is recommended that all message handlers observe the cancellation token to enable forced shutdown when required.
```

**Review by DavidBoike — APPROVED** [review]

**Pull request merged**